File: src/main/java/lechuck/intellij/util/RegexUtil.java
package lechuck.intellij.util;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Pattern;

public class RegexUtil {

    private RegexUtil() {}

    public static List<String> splitBySpacePreservingQuotes(String str) {
        if (str == null) {
            return List.of();
        }
        /* TODO Character classes should be preferred over reluctant quantifiers in regular expressions
            Found by SonarQube: Code smell Minor java:S5857
            .+? MUST be replaced by <[^>]++>
        */
        var matcher = Pattern.compile("([^\"]\\S*|\".+?\")\\s*").matcher(str);
        var result = new ArrayList<String>();
        while (matcher.find()) {
            String item = matcher.group(1).replace("\"", "");
            result.add(item);
        }
        return result;
    }
}

File: src/main/kotlin/lechuck/intellij/util/RegexUtil.kt
package lechuck.intellij.util

import java.util.regex.Pattern

object RegexUtil {
    fun splitBySpace(str: String?): List<String> {
        if (str == null) {
            return emptyList()
        }
        val matcher = Pattern.compile("([^\"]\\S*|\"[^\"]*\")\\s*").matcher(str)
        val result = ArrayList<String>()

        while (matcher.find()) {
            val item: String = matcher.group(1).replace("\"", "")
            result.add(item)
        }
        return result
    }
}